llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -7539,6 +7539,9 @@ VectorizationFactor LoopVectorizationPlanner::computeBestVF() {
                         CM.CostKind);
   precomputeCosts(BestPlan, BestFactor.Width, CostCtx);
   assert((BestFactor.Width == LegacyVF.Width ||
+          (BestPlan.getVectorLoopRegion() &&
+           BestPlan.getVectorLoopRegion()->getSingleSuccessor() !=
+               BestPlan.getMiddleBlock()) ||
```

Comment:
Might be good to pull this out together with an explanation.

Perhaps something like.
```suggestion
   // True if the BestPlan has been built from a loop with uncountable early exit. The legacy cost model doesn't properly model costs for such loops.
   bool PlanForEarlyExitLoop = BestPlan.getVectorLoopRegion() &&
           BestPlan.getVectorLoopRegion()->getSingleSuccessor() !=
               BestPlan.getMiddleBlock();
  assert((BestFactor.Width == LegacyVF.Width || PlanForEarlyExitLoop
           ||
```